### DISCUSSION TITLE
Adding typescript types declaration

### DISCUSSION THREAD

**Discussion by Mushr0000m** [alpinejs/alpine#1062] [4. General]
Great work with Alpine 👍 !

When using it and extracting some code in a function (instead of inline) typescript has issue with the type of the magic properties `$el`, `$refs`, etc. For example in this simple code : 

```
export function test() {
    return {
        foo: 'bar',

        lorem() {
            console.log(this.$el);
        },
    }
}
```
It create errors because TS doesn't know the types of `$el` and that it's available on `this`. Would it be possible to have some types declare to fix this ?

Thanks

**Comment by HugoDF** [comment]
I'm not sure that's going to be possible.

In essence, TypeScript is right, if that function is used outside of Alpine, there won't be `$el`.

If you've got an idea of what the typedefs would look like, feel free to open a PR.

**Comment by rbeard0330** [comment]
Could this be made to work with a judicious type assertion?  Like [this](https://www.typescriptlang.org/play?noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&alwaysStrict=false#code/PTAEBkEsCMCcENYE9QGMD2ATApgWAFCQB2ALtrAGbyragCCANgA7HYDC6Atk+kdqaADeBUKAAk2BgC5QACQAqAWXABRBtk78SI0CFDoSAC3L1mrUE1jom5EpGwBnAgF8CBDEQcl64AAoBJADkVAH0AIToAZRVQAF4hZ1B4B1MWPg5uXi0Abjd8PQBVBxMMHAJiMkpqWn9FJAyePgFsAA8yIkwUxjT2LkatIR0GdFgNGQAKAEo4gD5QADd0SEwXPIoAVyJUO15QTnq+rNIpmVqDzKbvYXxRUZJ12CJBm9FQADoPuj8g0IjogBodKJhqNOFNnq9Xh4HOh1G9hgBzcZGSAON4SBiTXIvUTOQEvZzY1z4IA)?  Of course, this won't save you if you call `myComponent()` yourself rather than letting Alpine do it, but it would give you working typechecking.

**Reply by HugoDF** [reply]
Yeah I think this is how the alpine-typescript project works -> https://github.com/LeanAdmin/alpine-typescript

**Reply by Mushr0000m** [reply]
Thanks for this tip ! I'll have a look, I didn't know the alpine-typescript project

**Reply by HugoDF** [reply]
It's quite recent I think 🙂
